### ISSUE TITLE
roughset classifier problem

### ISSUE THREAD

**Issue by alekin2** [fracpete/scikit-weka#9]
I have problem with roughset weka classifier:

from sklearn.calibration import CalibratedClassifierCV
from sklweka.classifiers import WekaEstimator
from sklweka.dataset import load_arff
jvm.start(packages=True)
X, y, meta = load_arff("C:\\Program Files\\Weka-3-9-6\\data\\iris.arff", class_index="last")
rough = WekaEstimator(classname="weka.classifiers.rules.RoughSet",
                         options=["-D","5","-R","0","-I","0","-X","3"])
roughfit=rough.fit(X,y)
roughfit.predict(X)
jvm.stop()

I get:

INFO:weka.core.jvm:JVM already running, call jvm.stop() first
Exception in thread "Thread-0" java.lang.IndexOutOfBoundsException: Index -1 out of bounds for length 3
	java.base/jdk.internal.util.Preconditions.outOfBounds(Unknown Source)
	java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Unknown Source)
	java.base/jdk.internal.util.Preconditions.checkIndex(Unknown Source)
	java.base/java.util.Objects.checkIndex(Unknown Source)
	java.base/java.util.ArrayList.get(Unknown Source)
	weka.core.Attribute.value(Attribute.java:778)
	weka.core.AbstractInstance.stringValue(AbstractInstance.java:668)
	weka.core.AbstractInstance.stringValue(AbstractInstance.java:644)
	rseslib.structure.data.formats.ArffDoubleDataInput.convertToDoubleData(ArffDoubleDataInput.java:219)
	weka.classifiers.AbstractRseslibClassifierWrapper.classifyInstance(AbstractRseslibClassifierWrapper.java:224)
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Unknown Source)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Unknown Source)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Unknown Source)
	at java.base/java.util.Objects.checkIndex(Unknown Source)
	at java.base/java.util.ArrayList.get(Unknown Source)
	at weka.core.Attribute.value(Attribute.java:778)
	at weka.core.AbstractInstance.stringValue(AbstractInstance.java:668)
	at weka.core.AbstractInstance.stringValue(AbstractInstance.java:644)
	at rseslib.structure.data.formats.ArffDoubleDataInput.convertToDoubleData(ArffDoubleDataInput.java:219)
	at weka.classifiers.AbstractRseslibClassifierWrapper.classifyInstance(AbstractRseslibClassifierWrapper.java:224)
Traceback (most recent call last):
  File "C:\ProgramData\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py", line 3460, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-26-0af5679b65e3>", line 9, in <module>
    roughfit.predict(X)
  File "C:\ProgramData\Anaconda3\lib\site-packages\sklweka\classifiers.py", line 135, in predict
    result.append(self.header_.class_attribute.value(int(self._classifier.classify_instance(inst))))
  File "C:\Users\Massimo\AppData\Roaming\Python\Python310\site-packages\weka\classifiers.py", line 128, in classify_instance
    return self._mc_classify(inst.jobject)
  File "C:\Users\Massimo\AppData\Roaming\Python\Python310\site-packages\javabridge\jutil.py", line 859, in fn
    raise JavaException(x)
javabridge.jutil.JavaException: Index -1 out of bounds for length 3

**Comment by fracpete** [comment]
What Java version are you using (`java -version`)? Try using Java 11.

**Comment by alekin2** [comment]
java -version
openjdk version "17.0.7" 2023-04-18
OpenJDK Runtime Environment Temurin-17.0.7+7 (build 17.0.7+7)
OpenJDK 64-Bit Server VM Temurin-17.0.7+7 (build 17.0.7+7, mixed mode, sharing)

**Comment by alekin2** [comment]
Other classifiers are ok.

from sklweka.classifiers import WekaEstimator
from sklweka.dataset import load_arff
import sklweka.jvm as jvm
jvm.start(packages=True)
X, y, meta = load_arff("C:\Program Files\Weka-3-8-6\data\iris.arff", class_index="last")
modello = WekaEstimator(classname="weka.classifiers.functions.Logistic",
options=["-R","1.0E-8","-M","-1","-num-decimal-places","4"])
modellofit=modello.fit(X,y)
ris=modellofit.predict(X)
print(ris)
jvm.stop()

is ok

**Comment by alekin2** [comment]
OK I've istalled java 11

java -version
openjdk version "11.0.19" 2023-04-18
OpenJDK Runtime Environment Temurin-11.0.19+7 (build 11.0.19+7)
OpenJDK 64-Bit Server VM Temurin-11.0.19+7 (build 11.0.19+7, mixed mode)

from sklweka.dataset import load_arff
import sklweka.jvm as jvm
jvm.start(packages=True)
X, y, meta = load_arff("C:\Program Files\Weka-3-8-6\data\iris.arff", class_index="last")
rough = WekaEstimator(classname="weka.classifiers.rules.RoughSet",
options=["-D","5","-R","0","-I","0","-X","3"])
roughfit=rough.fit(X,y)
res=roughfit.predict(X)
print(res)
jvm.stop()


Now I get:

DEBUG:weka.core.jvm:Adding bundled jars
DEBUG:weka.core.jvm:Classpath=['C:\\Users\\utente\\AppData\\Local\\anaconda3\\Lib\\site-packages\\javabridge\\jars\\rhino-1.7R4.jar', 'C:\\Users\\utente\\AppData\\Local\\anaconda3\\Lib\\site-packages\\javabridge\\jars\\runnablequeue.jar', 'C:\\Users\\utente\\AppData\\Local\\anaconda3\\Lib\\site-packages\\javabridge\\jars\\cpython.jar', 'C:\\Users\\utente\\AppData\\Local\\anaconda3\\lib\\site-packages\\weka\\lib\\arpack_combined.jar', 'C:\\Users\\utente\\AppData\\Local\\anaconda3\\lib\\site-packages\\weka\\lib\\core.jar', 'C:\\Users\\utente\\AppData\\Local\\anaconda3\\lib\\site-packages\\weka\\lib\\mtj.jar', 'C:\\Users\\utente\\AppData\\Local\\anaconda3\\lib\\site-packages\\weka\\lib\\python-weka-wrapper.jar', 'C:\\Users\\utente\\AppData\\Local\\anaconda3\\lib\\site-packages\\weka\\lib\\weka.jar']
DEBUG:weka.core.jvm:MaxHeapSize=default
DEBUG:weka.core.jvm:Package support enabled
lug 03, 2023 10:52:14 AM com.github.fommil.netlib.ARPACK <clinit>
AVVERTENZA: Failed to load implementation from: com.github.fommil.netlib.NativeSystemARPACK
lug 03, 2023 10:52:14 AM com.github.fommil.netlib.ARPACK <clinit>
AVVERTENZA: Failed to load implementation from: com.github.fommil.netlib.NativeRefARPACK
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by weka.core.WekaPackageClassLoaderManager (file:/C:/Users/utente/AppData/Local/anaconda3/Lib/site-packages/weka/lib/weka.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
WARNING: Please consider reporting this to the maintainers of weka.core.WekaPackageClassLoaderManager
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
[WekaPackageManager] Skipping package timeseriesForecasting because it is not compatible with Weka 3.9.6
Exception in thread "Thread-0" java.lang.IndexOutOfBoundsException: Index -1 out of bounds for length 3
	java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64)
	java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70)
	java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:248)
	java.base/java.util.Objects.checkIndex(Objects.java:372)
	java.base/java.util.ArrayList.get(ArrayList.java:459)
	weka.core.Attribute.value(Attribute.java:778)
	weka.core.AbstractInstance.stringValue(AbstractInstance.java:668)
	weka.core.AbstractInstance.stringValue(AbstractInstance.java:644)
	rseslib.structure.data.formats.ArffDoubleDataInput.convertToDoubleData(ArffDoubleDataInput.java:219)
	weka.classifiers.AbstractRseslibClassifierWrapper.classifyInstance(AbstractRseslibClassifierWrapper.java:224)
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:248)
	at java.base/java.util.Objects.checkIndex(Objects.java:372)
	at java.base/java.util.ArrayList.get(ArrayList.java:459)
	at weka.core.Attribute.value(Attribute.java:778)
	at weka.core.AbstractInstance.stringValue(AbstractInstance.java:668)
	at weka.core.AbstractInstance.stringValue(AbstractInstance.java:644)
	at rseslib.structure.data.formats.ArffDoubleDataInput.convertToDoubleData(ArffDoubleDataInput.java:219)
	at weka.classifiers.AbstractRseslibClassifierWrapper.classifyInstance(AbstractRseslibClassifierWrapper.java:224)
Traceback (most recent call last):
  File "C:\Users\utente\AppData\Local\anaconda3\lib\site-packages\IPython\core\interactiveshell.py", line 3460, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-2-d6859873cda8>", line 9, in <module>
    res=roughfit.predict(X)
  File "C:\Users\utente\AppData\Local\anaconda3\lib\site-packages\sklweka\classifiers.py", line 135, in predict
    result.append(self.header_.class_attribute.value(int(self._classifier.classify_instance(inst))))
  File "C:\Users\utente\AppData\Local\anaconda3\lib\site-packages\weka\classifiers.py", line 128, in classify_instance
    return self._mc_classify(inst.jobject)
  File "C:\Users\utente\AppData\Local\anaconda3\lib\site-packages\javabridge\jutil.py", line 859, in fn
    raise JavaException(x)
javabridge.jutil.JavaException: Index -1 out of bounds for length 3

**Comment by fracpete** [comment]
What Weka package did you install?

**Comment by alekin2** [comment]
I tried on 2 pc, weka 3.8.6 and weka 3.9.6, only roughset get problem.
however, I installed more packages than the basic ones. Here is the list from weka package manager:

AffectiveTweets	Text classification	1.0.2	1.0.2	Yes
alternatingDecisionTrees	Classification	1.0.7	1.0.7	Yes
alternatingModelTrees	Regression	1.0.1	1.0.1	Yes
AnalogicalModeling	Classification	0.12.0	0.12.0	Yes
AnDE	Classification	1.2.1	1.2.1	Yes
attributeSelectionSearchMethods	Attribute selection	1.0.7	1.0.7	Yes
Auto-WEKA	Classification, Regression, Attribute Selection	2.6.4	2.6.4	Yes
BANGFile	Clustering	1.0.0	1.0.0	Yes
bayesianLogisticRegression	Text classification	1.0.5	1.0.5	Yes
bestFirstTree	Classification	1.0.4	1.0.4	Yes
BESTrees	Ensemble learning, Classification, Regression	1.0.1	-----	Yes
CAIRAD	Classification	1.0	1.0	Yes
calibrationCurve	Visualization	1.0.0	1.0.0	Yes
cascadeKMeans	Clustering	1.0.4	1.0.4	Yes
CFWNB	Classification	1.0.0	1.0.0	Yes
CHIRP	Classification	1.0.1	1.0.1	Yes
chiSquaredAttributeEval	Attribute selection	1.0.4	1.0.4	Yes
citationKNN	Multi-instance learning	1.0.1	1.0.1	Yes
classificationViaClustering	Classification	1.0.8	1.0.8	Yes
classifierBasedAttributeSelection	Attribute selection	1.0.5	1.0.5	Yes
clojureClassifier	Classification	1.0.1	1.0.1	Yes
CLOPE	Clustering	1.0.1	1.0.1	Yes
collective-classification	Classification	2019.8.7	-----	Yes
complementNaiveBayes	Classification	1.0.3	1.0.3	Yes
conjunctiveRule	Classification	1.0.4	1.0.4	Yes
consistencySubsetEval	Attribute selection	1.0.5	1.0.5	Yes
costSensitiveAttributeSelection	Attribute selection	1.0.3	1.0.3	Yes
CSForest	Classification	1.1	1.1	Yes
CVAttributeEval	Attribute selection	1.0.0	1.0.0	Yes
dagging	Ensemble learning	1.0.3	1.0.3	Yes
decorate	Ensemble learning	1.0.2	1.0.2	Yes
denormalize	Preprocessing	1.0.3	1.0.3	Yes
DilcaDistance	Distance 	1.0.2	1.0.2	Yes
discriminantAnalysis	Classification	1.0.4	1.0.4	Yes
DistributionBasedBalance	Preprocessing	1.0.0	1.0.0	Yes
DMNBtext	Text classification	1.0.1	1.0.1	Yes
DTNB	Classification	1.0.4	1.0.4	Yes
dualPerturbAndCombine	Classification and regression	1.0.0	1.0.0	Yes
EBMC	Classification	1.0.0	1.0.0	Yes
elasticNet	Regression	1.0.1	1.0.1	Yes
ensembleLibrary	Ensemble learning	1.0.6	1.0.6	Yes
ensemblesOfNestedDichotomies	Ensemble learning	1.0.6	1.0.6	Yes
EvolutionarySearch	Attribute selection	1.0.2	1.0.2	Yes
extraTrees	Classification	1.0.2	1.0.2	Yes
fastCorrBasedFS	Attribute selection	1.0.1	1.0.1	Yes
filteredAttributeSelection	Attribute selection	1.0.1	1.0.1	Yes
ForestPA	Classification	1.0	1.0	Yes
ForExPlusPlus	Classification	1.1	1.1	Yes
fourierTransform	Preprocessing	1.0.1	1.0.1	Yes
functionalTrees	Classification	1.0.4	1.0.4	Yes
fuzzyLaticeReasoning	Classification	1.0.1	1.0.1	Yes
fuzzyUnorderedRuleInduction	Classification	1.0.2	1.0.2	Yes
GenClustPlusPlus	Clustering	1.5	1.5	Yes
generalizedSequentialPatterns	Associations	1.0.1	1.0.1	Yes
GPAttributeGeneration	Classification, Preprocessing	1.0.1	1.0.1	Yes
grading	Ensemble learning	1.0.1	1.0.1	Yes
gridSearch	Classification	1.0.12	1.0.12	Yes
hiddenNaiveBayes	Classification	1.0.1	1.0.1	Yes
HMM	Classification, Multiinstance, Sequence	0.1.1	0.1.1	Yes
hyperPipes	Classification	1.0.1	1.0.1	Yes
IBkLG	Classification	1.0.1	1.0.1	Yes
isolationForest	Outlier	1.0.2	1.0.2	Yes
isotonicRegression	Regression	1.0.1	1.0.1	Yes
iterativeAbsoluteErrorRegression	Regression	1.0.0	1.0.0	Yes
IWSS	Attribute selection	1.0.0	1.0.0	Yes
IWSSembeddedNB	Attribute selection	1.0.0	1.0.0	Yes
J48Consolidated	Classification	3.2	3.2	Yes
JCDT	Classification, Regression	1.0.0	1.0.0	Yes
JCHAIDStar	Classification	1.2	1.2	Yes
kerasZoo	Python integration	1.0.3	1.0.3	Yes
kernelLogisticRegression	Classification	1.0.0	1.0.0	Yes
kfGroovy	KnowledgeFlow	1.0.15	1.0.16	Yes
largeScaleKernelLearning	Preprocessing	1.0.4	1.0.4	Yes
lazyAssociativeClassifier	Classification	1.0.0	1.0.0	Yes
lazyBayesianRules	Classification	1.0.1	1.0.1	Yes
leastMedSquared	Regression	1.0.1	1.0.1	Yes
levenshteinEditDistance	Distance measure	1.0.1	1.0.1	Yes
LibLINEAR	Classification	1.9.8	1.9.8	Yes
LibSVM	Classification, Regression	1.0.10	1.0.10	Yes
linearForwardSelection	Attribute selection	1.0.1	1.0.1	Yes
logarithmicErrorMetrics	Metrics	1.0.0	1.0.0	Yes
LVQ	Clustering	1.0.1	1.0.1	Yes
massiveOnlineAnalysis	Data streams	2023.04.0	2023.04.0	Yes
metaCost	Classification	1.0.3	1.0.3	Yes
metaphorSearchMethods	Attribute selection	1.0.2	1.0.2	Yes
MODLEM	Classification, Ensemble learning	1.0.0	1.0.0	Yes
multiBoostAB	Ensemble learning	1.0.1	1.0.1	Yes
multiInstanceFilters	Preprocessing	1.0.10	1.0.10	Yes
multiInstanceLearning	Multi-instance learning	1.0.10	1.0.10	Yes
multilayerPerceptronCS	Classification	1.0.1	1.0.1	Yes
multiLayerPerceptrons	Classification/regression, Preprocessing	1.0.10	1.0.10	Yes
MultiObjectiveEvolutionaryFuzzyClassifier	Classification	1.0.1	1.0.1	Yes
MultiObjectiveEvolutionarySearch	Attribute selection	1.0.0	1.0.0	Yes
multisearch	Classification	2020.2.17	2020.2.17	Yes
mxexpression	Preprocessing	2018.3.2	-----	Yes
naiveBayesTree	Classification	1.0.4	1.0.4	Yes
netlibNativeWindows	Linear Algebra	1.0.1	1.0.1	Yes
NeuralNetwork	Classification, Regression	1.0	-----	Yes
NNge	Classification	1.0.3	1.0.3	Yes
normalize	Preprocessing	1.0.1	1.0.1	Yes
oneClassClassifier	Classification	1.0.4	1.0.4	Yes
optics_dbScan	Clustering	1.0.6	1.0.6	Yes
OptimizedForest	Classification	1.0.2	1.0.2	Yes
ordinalClassClassifier	Classification	1.0.5	1.0.5	Yes
ordinalLearningMethod	Classification	1.0.1	1.0.1	Yes
ordinalStochasticDominance	Classification	1.0.1	1.0.1	Yes
paceRegression	Regression	1.0.1	1.0.1	Yes
PairwiseConsistencyAttributeEval	Attribute selection	1.0.1	1.0.1	Yes
PairwiseCorrelationAttributeEval	Attribute selection	1.0.1	1.0.1	Yes
partialLeastSquares	Preprocessing	1.0.5	1.0.5	Yes
percentageErrorMetrics	Metrics	1.0.1	1.0.1	Yes
probabilisticSignificanceAE	Attribute Selection	1.0.1	1.0.1	Yes
probabilityCalibrationTrees	Probability calibration, Ensemble learning	1.0.0	1.0.0	Yes
PSOSearch	Attribute selection	1.2.0	1.2.0	Yes
racedIncrementalLogitBoost	Ensemble learning	1.0.1	1.0.1	Yes
raceSearch	Attribute Selection	1.0.1	1.0.1	Yes
RankCorrelation	Metrics	1.0.0	1.0.0	Yes
RankerByDTClassification	Classification, Visualization	1.0.0	1.0.0	Yes
RBFNetwork	Classification/regression	1.0.8	1.0.8	Yes
realAdaBoost	Ensemble learning	1.0.1	1.0.1	Yes
RerankingSearch	Attribute selection	1.1.0	1.1.0	Yes
ridor	Classification	1.0.1	1.0.1	Yes
rotationForest	Ensemble learning	1.0.2	1.0.2	Yes
Rseslib	Classification	3.3.0	3.3.0	Yes
SelfOrganizingMap	Clustering	1.0.3	1.0.3	Yes
sequentialInformationalBottleneckClusterer	Clustering	1.0.1	1.0.1	Yes
simpleCART	Classification	1.0.4	1.0.4	Yes
simpleEducationalLearningSchemes	Classification	1.0.1	1.0.1	Yes
SmoothPrivateForest	Classification	1.2.1	1.2.1	Yes
SMOTE	Preprocessing	1.0.3	1.0.3	Yes
SPAARC	Classification	1.0	1.0	Yes
SparseGenerativeModel	Text classification	1.4.9	1.4.9	Yes
SPegasos	Classification	1.0.1	1.0.1	Yes
SPMFWrapper	Associations	0.0.2	0.0.2	Yes
stackingC	Ensemble learning	1.0.4	1.0.4	Yes
StudentFilters	Preprocessing	2.0.0	2.0.0	Yes
supervisedAttributeScaling	Preprocessing	1.0.2	1.0.2	Yes
SVMAttributeEval	Attribute selection	1.0.1	1.0.1	Yes
SysFor	Classification	1.2	1.2	Yes
tabuAndScatterSearch	Attribute selection	1.0.1	1.0.1	Yes
tertius	Associations	1.0.1	1.0.1	Yes
thresholdSelector	Classification	1.0.3	1.0.3	Yes
tigerjython	Scripting	1.0.5	1.0.5	Yes
ultraBoost	Classification	1.0.0	1.0.0	Yes
userClassifier	Classification/regression	1.0.3	1.0.3	Yes
vines	Density Estimation	1.0.0	1.0.0	Yes
votingFeatureIntervals	Classification	1.0.1	1.0.1	Yes
wekaclassalgos	Classification	2017.10.18	-----	Yes
wekaDeeplearning4j	Classification/Regression	1.7.2	1.7.2	Yes
WekaExcel	Converter	1.0.8	1.0.8	Yes
WekaPyScript	Classification	0.6.0	0.6.0	Yes
wekaPython	Python integration	1.0.13	1.0.17	Yes
winnow	Classification	1.0.1	1.0.1	Yes
WiSARD	Classification	1.0.1	1.0.1	Yes
XMeans	Clustering	1.0.6	1.0.6	Yes

**Comment by fracpete** [comment]
Which Weka package contains the class `weka.classifiers.rules.RoughSet`? It's not part of the core Weka distribution, hence I need to know what Weka package this is from so I can test it (I'm not gonna install all the packages that you've listed).

**Comment by alekin2** [comment]
Try https://weka.sourceforge.io/packageMetaData/Rseslib/Latest.html

**Comment by fracpete** [comment]
Thanks. I had to implement a minor fix (32b43048b0ce1e0d928f19ebf362d3ace18418c1).

The following code now works with the just released version 0.0.7:

```python
from sklweka.dataset import load_arff
from sklweka.classifiers import WekaEstimator
import sklweka.jvm as jvm

jvm.start(packages=True)

X, y, meta = load_arff("/home/fracpete/development/datasets/uci/iris.arff", class_index="last")
rough = WekaEstimator(classname="weka.classifiers.rules.RoughSet",
                      options=["-D", "5", "-R", "0", "-I", "0", "-X", "3"])
roughfit = rough.fit(X,y)
res = roughfit.predict(X)
print(res)

jvm.stop()
```
